src/NuGet.Core/NuGet.Packaging/PackageArchiveReader.cs

```diff
@@ -296,7 +296,10 @@ public override Task<bool> IsSignedAsync(CancellationToken token)
             {
                 var hashAlgorithm = signatureManifest.HashAlgorithm.GetHashProvider();
                 var expectedHash = Convert.FromBase64String(signatureManifest.HashValue);
-                SignedPackageArchiveUtility.VerifySignedZipIntegrity(reader, hashAlgorithm, expectedHash);
+                if (!SignedPackageArchiveUtility.VerifySignedZipIntegrity(reader, hashAlgorithm, expectedHash))
+                {
+                    throw new SignatureException(Strings.SignatureFailurePackageTampered);
```

Comment:
Include the path if possible. This could come from checking if the stream is a FileStream and casting it to get the path. During a large restore with hundreds of packages it can be really confusing if there is a generic error message that doesn't specify the source.